httpstan/stan_services.cpp

```diff
@@ -164,6 +164,34 @@ double log_prob(py::dict data,
     std::rethrow_exception(p);
 }
 
+// See exported docstring
+std::vector<double> log_prob_grad(py::dict data,
+                                  const std::vector<double>& unconstrained_parameters,
+                                  bool adjust_transform) {
+  stan::io::array_var_context &var_context = new_array_var_context(data);
+  // random_seed, the second argument, is unused but the function requires it.
+  stan::model::model_base &model = new_model(var_context, (unsigned int)1, &std::cout);
+  if(unconstrained_parameters.size() != model.num_params_r()) {
+    throw std::runtime_error("The number of parameters does not match the number of unconstrained parameters in the model.");
+  }
+  // copy unconstrained_parameters -> params_r
+  std::vector<double> params_r(unconstrained_parameters.size());
+  std::memcpy(params_r.data(), unconstrained_parameters.data(), unconstrained_parameters.size()*sizeof(double));
+  // calculate logprob
+  std::vector<int> params_i(model.num_params_i(), 0);
```

Comment:
Two questions:
1. Is the copy into params_r necessary?
2. What about params_i? What if the model has integer parameters?

Reply:
responses below

> 1. Is the copy into params_r necessary?
If the unconstrained_parameters argument is defined as `const std::vector<double>& unconstrained_parameters` then the copy to params_r is necessary, defining as `std::vector<double> unconstrained_parameters` means the copy isn't necessary. I'm not sure which is preferred?

> 2. What about params_i? What if the model has integer parameters?
The model base class (see https://github.com/stan-dev/stan/blob/develop/src/stan/model/model_base.hpp) suggests that `params_i` are ignored, so I don't think it's necessary to have an argument for them.